ocaml/xapi/message_forwarding.ml

```diff
@@ -1694,7 +1694,20 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
             fun ~local_fn ~__context ~vm op ->
               finally
                 (fun () -> allocate_vm_to_host ~__context ~vm ~host ~snapshot ~host_op:`vm_migrate ();
-                           forward_vm_op ~local_fn ~__context ~vm op)
+                           let internal_async_op ~__context ~wait =
```

Comment:
I've been looking at this patch for a while now, and have to admit that I don't really follow what's going on anymore.

Part of the problem is that the actual API calls that are being executed (`{Local;Client;Client.InternalAsync}.VM.migrate_send`) are at three different places inside the function `migrate_send`. Furthermore, we have `forward_vm_op`, which inside it calls `do_op_on`, but there is also a call to `do_op_on_no_localhost` used for just the internal-async call.

The structure of the function was already quite complicated, because it includes three cases: non-live migration, live intra-pool migration, and live cross-pool migration. I believe that you have now "treated" just the live intra-pool case with internal-async logic, while at least the cross-pool case would also benefit from it.

Unfortunately, we have ended up with a mix of forwarding logic and different API calls that is hard to reason about.

Reply:
have you tried diffing against the commit before the original patch? that might make it slightly easier to see what's going on

i will try to make it more explicit which migration case we're in, perhaps that would make it easier as well

i will look into adding functionality for the cross-pool case. admittedly i haven't been testing with the stunnel idle timeout set to 60s because i've just been trying to get it to work - this should flag any problems with the cross-pool case

Reply:
Yep, I am familiar with the original code ;)

Reply:
i also got rid of some unnecessary function args (e.g. __context, local_fn etc.), since these are all in scope anyway